### PR TITLE
[WIP] Implement the thing

### PR DESCRIPTION
**Pull request by magik6k**

### CODE REVIEW

**Comment by daviddias** [comment]
👏🏽👏🏽👏🏽👏🏽👏🏽👏🏽👏🏽👏🏽👏🏽👏🏽 @magik6k :D 

Does this work in [`ipld-resolver`](https://github.com/ipld/js-ipld-resolver) with tests that use actual git objects?

**Comment by magik6k** [comment]
@diasdavid done in https://github.com/ipld/js-ipld-resolver/pull/97, the tests there pass locally for me, though this needs to be reviewed/merged/released first.

File: src/util.js [hunk 1]
```diff
@@ -1,19 +1,77 @@
...
+      tree.serialize(dagNode, callback)
+  }
+}
+
+exports.deserialize = (data, callback) => {
+  let headLen = gitUtil.find(data, 0)
+  let head = data.slice(0, headLen).toString()
+  let typeLen = head.match(/([^ ]+) (\d+)/)
+  if (!typeLen) {
+    setImmediate(() => callback(new Error('invalid object header'), null))
```
Comment (daviddias) [inline src/util.js:50]: You gotta say `return` otherwise the rest of the function will be executed.

File: src/resolver.js [hunk 2]
```diff
@@ -1,16 +1,152 @@
...
+      case 'tag':
+        paths = [
+          'object',
+          'type',
+          'tag',
+          'message'
+        ]
+
+        if (node.tagger) {
+          paths = paths.concat(personInfoPaths.map(e => 'tagger/' + e))
```
Comment (daviddias) [inline src/resolver.js]: Add the parenthesis to the arrow function: `.map((e) => 'tagger/' + e))

File: src/resolver.js [hunk 3]
```diff
@@ -1,16 +1,152 @@
...
+          paths.push('encoding')
+        }
+        break
+      case 'tag':
+        paths = [
+          'object',
+          'type',
+          'tag',
+          'message'
+        ]
```
Comment (daviddias) [inline src/resolver.js:118]: Are these constants? Move to a Constants file

File: src/resolver.js [hunk 4]
```diff
@@ -1,16 +1,152 @@
...
 
-exports.multicodec = 'git'
+exports.multicodec = 'git-raw'
+
+const personInfoPaths = [
+  'original',
+  'name',
+  'email',
+  'date'
+]
```
Comment (daviddias) [inline src/resolver.js:15]: Are these constants? Move to a constant file

File: package.json [hunk 5]
```diff
@@ -37,10 +37,13 @@
...
-    "cids": "~0.5.0",
-    "multihashes": "~0.4.5",
-    "multihashing-async": "~0.4.5",
+    "async": "^2.5.0",
+    "cids": "^0.5.1",
+    "multicodec": "^0.1.9",
+    "multihashes": "^0.4.8",
+    "multihashing-async": "^0.4.6",
+    "smart-buffer": "^3.0.3",
+    "traverse": "^0.6.6"
```
Comment (daviddias) [inline package.json]: Replace the `^` for `~` for any module that is below 1.0.0

**Review by daviddias — CHANGES_REQUESTED** [review]
There is a ton of good work here! Really impressed and I would not say that you are not used to JS :) 

Added some comments for small changes. Otherwise LGTM

File: src/util.js (hunk 1 above)
Comment (magik6k) [reply src/util.js:50]: done

File: src/resolver.js (hunk 3 above)
Comment (magik6k) [reply src/resolver.js:118]: These few paths are there for every tag, I'm not really sure if it'd really be worth having a separate file for that.

**Review by daviddias — APPROVED** [review]

**Pull request merged**